src/org/traccar/protocol/MiniFinderProtocolDecoder.java

```diff
@@ -34,44 +34,145 @@ public MiniFinderProtocolDecoder(MiniFinderProtocol protocol) {
+            .number("(-?d+.d+),")                // altitude (meters)
+            .number("(d+),")                     // battery (percentage)
             .number("(d+),")                     // satellites in use
             .number("(d+),")                     // satellites in view
-            .text("0")
-            .or()
-            .any()
-            .groupEnd()
+            .number("(d+.?d*)")                  // HDOP, confirmed with Eview manufacturer
             .compile();
 
+    private void decodeFlags(Position position, int flags) {
+
+        position.setValid(BitUtil.check(flags, 0));
+
+        if (BitUtil.check(flags, 2)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_FAULT);
+        }
+        if (BitUtil.check(flags, 6)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_SOS);
+        }
+        if (BitUtil.check(flags, 7)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_OVERSPEED);
+        }
+        if (BitUtil.check(flags, 8)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_FALL_DOWN);
+        }
+        if (BitUtil.check(flags, 9) || BitUtil.check(flags, 10) || BitUtil.check(flags, 11)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_GEOFENCE);
+        }
+        if (BitUtil.check(flags, 12)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_LOW_BATTERY);
+        }
+        if (BitUtil.check(flags, 15) || BitUtil.check(flags, 14)) {
+            position.set(Position.KEY_ALARM, Position.ALARM_MOVEMENT);
+        }
+
+        position.set(Position.KEY_RSSI, BitUtil.between(flags, 16, 20));
+        position.set(Position.KEY_CHARGE, BitUtil.check(flags, 22));
+    }
+
     @Override
     protected Object decode(
             Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
 
         String sentence = (String) msg;
 
-        if (sentence.startsWith("!1")) {
+        if (sentence.startsWith("!1,")) {
 
             getDeviceSession(channel, remoteAddress, sentence.substring(3, sentence.length()));
 
-        } else if (sentence.matches("![A-D].*")) {
+            return null;
+        }
 
-            DeviceSession deviceSession = getDeviceSession(channel, remoteAddress);
-            if (deviceSession == null) {
+        DeviceSession deviceSession = getDeviceSession(channel, remoteAddress);
+        if (deviceSession == null) {
+            return null;
+        }
+
+        if (sentence.startsWith("!B,") || sentence.startsWith("!D,")) {
+            Parser parser = new Parser(PATTERN_BD, sentence);
+            if (!parser.matches()) {
                 return null;
             }
 
-            Parser parser = new Parser(PATTERN, sentence);
+            Position position = new Position();
+            position.setProtocol(getProtocolName());
+            position.setDeviceId(deviceSession.getDeviceId());
+
+            DateBuilder dateBuilder = new DateBuilder()
```

Comment:
In my initial attempt, I felt the code became too cumbersome and un-intuitive.  I will try again.